src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,694 @@
+    }
+
+    return table_is_valid;
+}
+
+Chunk StorageKeeperMap::getByKeys(const ColumnsWithTypeAndName & keys, PaddedPODArray<UInt8> & null_map, const Names &) const
+{
+    if (keys.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "StorageKeeperMap supports only one key, got: {}", keys.size());
+
+    auto raw_keys = serializeKeysToRawString(keys[0]);
+
+    if (raw_keys.size() != keys[0].column->size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Assertion failed: {} != {}", raw_keys.size(), keys[0].column->size());
+
+    return getBySerializedKeys(raw_keys, &null_map);
+}
+
+Chunk StorageKeeperMap::getBySerializedKeys(const std::span<const std::string> keys, PaddedPODArray<UInt8> * null_map) const
+{
+    Block sample_block = getInMemoryMetadataPtr()->getSampleBlock();
+    MutableColumns columns = sample_block.cloneEmptyColumns();
+    size_t primary_key_pos = getPrimaryKeyPos(sample_block, getPrimaryKey());
+
+    if (null_map)
+    {
+        null_map->clear();
+        null_map->resize_fill(keys.size(), 1);
+    }
+
+    auto client = getClient();
+
+    std::vector<std::future<Coordination::GetResponse>> values;
+    values.reserve(keys.size());
+
+    for (const auto & key : keys)
+    {
+        const auto full_path = fullPathForKey(key);
+        if (full_path == metadata_path)
+        {
+            values.emplace_back();
+            continue;
+        }
+
+        values.emplace_back(client->asyncTryGet(full_path));
+    }
+
+    auto wait_until = std::chrono::system_clock::now() + std::chrono::milliseconds(Coordination::DEFAULT_OPERATION_TIMEOUT_MS);
+
+    for (size_t i = 0; i < keys.size(); ++i)
+    {
+        auto & value = values[i];
+        if (!value.valid())
+            continue;
+
+        if (value.wait_until(wait_until) != std::future_status::ready)
+            throw DB::Exception(ErrorCodes::KEEPER_EXCEPTION, "Failed to fetch values: timeout");
+
+        auto response = value.get();
+        Coordination::Error code = response.error;
+
+        if (code == Coordination::Error::ZOK)
+        {
+            fillColumns(base64Decode(keys[i], true), response.data, primary_key_pos, sample_block, columns);
+        }
+        else if (code == Coordination::Error::ZNONODE)
+        {
+            if (null_map)
+            {
+                (*null_map)[i] = 0;
+                for (size_t col_idx = 0; col_idx < sample_block.columns(); ++col_idx)
+                    columns[col_idx]->insert(sample_block.getByPosition(col_idx).type->getDefault());
+            }
+        }
+        else
+        {
+            throw DB::Exception(ErrorCodes::KEEPER_EXCEPTION, "Failed to fetch value: {}", code);
+        }
+    }
+
+    size_t num_rows = columns.at(0)->size();
+    return Chunk(std::move(columns), num_rows);
+}
+
+Block StorageKeeperMap::getSampleBlock(const Names &) const
+{
+    auto metadata = getInMemoryMetadataPtr();
+    return metadata ? metadata->getSampleBlock() : Block();
```

Comment:
When is it false?

Reply:
Not sure, I added it to be aligned with some new `EmbeddedRocksDB` changes (the function is declared in `IKeyValueEntity`). 
If it can't be `nullptr` I'll remove the check.

Reply:
Seems like it can't